osquery/dispatcher/tests/task_schedule_tests.cpp

```diff
@@ -0,0 +1,145 @@
+  }
+
+  void TearDown() override {
+    FLAGS_disable_logging = logging_;
+  }
+
+ private:
+  bool logging_{false};
+};
+
+TEST_F(TaskScheduleTests, run_for_the_one_time) {
+  auto tasks = TaskSchedule{};
+  auto counter = int{0};
+  tasks.add(
+      [&counter](const auto scheduled_time) {
+        ++counter;
+        return 0;
+      },
+      0);
+  while (!tasks.isEmpty()) {
+    auto next_time_to_run = tasks.nextTimeToRun();
+    auto now = getUnixTime();
+    if (now < next_time_to_run) {
+      std::this_thread::sleep_for(std::chrono::seconds{next_time_to_run - now});
+    } else {
+      tasks.runNextNow();
+    }
+  }
+  EXPECT_EQ(counter, 1);
+}
+
+TEST_F(TaskScheduleTests, first_start_priority) {
+  auto tasks = TaskSchedule{};
+  auto counter = int{0};
+  tasks.add(
+      [&counter](const auto scheduled_time) {
+        ++counter;
+        return 0;
+      },
+      10);
+  tasks.add(
+      [&counter](const auto scheduled_time) {
+        counter *= 2;
+        return 0;
+      },
+      100);
+  while (!tasks.isEmpty()) {
+    auto next_time_to_run = tasks.nextTimeToRun();
+    auto now = getUnixTime();
+    if (now < next_time_to_run) {
+      std::this_thread::sleep_for(std::chrono::seconds{next_time_to_run - now});
+    } else {
+      tasks.runNextNow();
+    }
+  }
+  EXPECT_EQ(counter, 2);
+}
+
+TEST_F(TaskScheduleTests, one_shot_task) {
+  auto tasks = TaskSchedule{};
+  auto one_shot_counter = int{0};
+  auto counter = int{0};
+  tasks.add([&counter](const auto scheduled_time) {
+    ++counter;
+    return 1;
+  });
+  tasks.add([&one_shot_counter](const auto scheduled_time) {
+    ++one_shot_counter;
+    return 0;
+  });
+  for (int i = 0; i < 10 && !tasks.isEmpty(); ++i) {
+    auto next_time_to_run = tasks.nextTimeToRun();
+    auto now = getUnixTime();
+    if (now < next_time_to_run) {
+      std::this_thread::sleep_for(std::chrono::seconds{next_time_to_run - now});
+    } else {
+      tasks.runNextNow();
+    }
+  }
+  EXPECT_EQ(one_shot_counter, 1);
+  EXPECT_EQ(counter, 9);
+}
+
+TEST_F(TaskScheduleTests, zero_time_to_wait) {
+  auto tasks = TaskSchedule{};
+  tasks.add([](const auto scheduled_time) { return scheduled_time; });
+  auto sum_waiting_time = decltype(tasks.nextTimeToRun()){0};
+  for (int i = 0; i < 10 && !tasks.isEmpty(); ++i) {
+    auto next_time_to_run = tasks.nextTimeToRun();
+    auto now = getUnixTime();
+    if (now < next_time_to_run) {
+      auto time_to_wait = next_time_to_run - now;
+      sum_waiting_time += time_to_wait;
+      std::this_thread::sleep_for(std::chrono::seconds{time_to_wait});
+    } else {
+      tasks.runNextNow();
+    }
+  }
+  EXPECT_EQ(sum_waiting_time, 0u);
+}
+
+TEST_F(TaskScheduleTests, run_next_regardless_of_time) {
+  auto tasks = TaskSchedule{};
+  auto counter = int{0};
+  tasks.add([&counter](const auto scheduled_time) {
+    ++counter;
+    return scheduled_time + 10;
+  });
+  for (int i = 0; i < 10 && !tasks.isEmpty(); ++i) {
```

Comment:
Make sure the values are what you want on every cycle, instead of checking `tasks.isEmpty()` on loop conditions, just make ensure you test it with `EXPECT_EQ` on the body of the loop. Same goes for `counter` on every iteration etc.